### ISSUE TITLE
Jest compatibility

### ISSUE THREAD

**Issue by t1bb4r** [analog-nico/stealthy-require#5]
[Jest](https://github.com/facebook/jest) is a great test runner and I'm trying to use it with [request-promise](https://github.com/request/request-promise-native) in my current project. The problem is this module isn't working with jest.

Here is a minimal example:
`reusable.js`:
```
let id = Math.random();
module.exports = {id};
```
`users.js`:
```
let stealthyRequire = require('stealthy-require');
var reusableStealth = stealthyRequire(require.cache, function () {
    return require('./reusable');
},
function () {}, module);
var reusable = require("./reusable");
module.exports = {
  reusableStealth,
  reusable
}
```
`__tests__/users.js`:
```
let users = require("../users");
describe("", () => {
  it("", () => {
    expect(users.reusableStealth.id).not.toEqual(users.reusable.id);
  })
})
```

This actually works in node, when I require the users module I get two unique ids. Just not working with jest, so I can understand if you don't consider this a bug.

Side-note: I don't understand how I can be the only one using jest and request, each have over 10 000 stars on github.

**Comment by t1bb4r** [comment]
Doesn't look like there is anything wrong with this module, it's more of a request-promise-core / jest issue.

EDIT: In jest `require.cache` doesn't actually control the cache, so request-promise-core and jest is incompatible)

**Comment by analog-nico** [comment]
Hey @t1bb4r , I am curious nonetheless. Since the code is working in node without Jest I am wondering what difference Jest makes. In your Jest version do you still execute it within node or somewhere else like in the browser? Do you use transpilers like babel? And packagers like webpack or browserify?

**Comment by t1bb4r** [comment]
I'm using `jest --env=node` which should execute the tests in a node environment. They have their own module cache / registry management, which is the problem (I might be completely wrong). I've been investigating this file, [jest-runtime](https://github.com/facebook/jest/blob/master/packages/jest-runtime/src/index.js), for a while to try and understand whats going on.

Jest has a mocking feature where you can mock any import, this feature is probably the reason they keep their own module registry.

**Comment by analog-nico** [comment]
Alright, I see. They are [passing their own `require` function](https://github.com/facebook/jest/blob/master/packages/jest-runtime/src/index.js#L519-L528) to the modules and that function [uses a cache that is not exposed](https://github.com/facebook/jest/blob/master/packages/jest-runtime/src/index.js#L284). So yes, `stealthy-require` and `jest` are incompatible and there is no solution unless `jest` would expose their `moduleRegistry`.

But thinking outside the box: You want to test your `request-promise` code, right? It uses `stealthy-require` internally to allow projects that use `request` and `request-promise` at the same time. If your project always uses `request-promise` and never `request` by itself then you can just ignore the fact that `stealthy-require` is falling back to a standard `require`.

**Comment by t1bb4r** [comment]
Its tough to say exactly what all the libraries I'm including uses internally and when they will conflict, but you are right, if we can manage to only use the core module once everything is ok.

The first time I started to run into this issue was when I installed a module which depends on "request-promise", while I'm using "request-promise-native". These imports, no matter how deep in your dependency tree, will break jest.

Either way, thanks a lot for the help and feedback. I'll manage to workaround the issue.
